README.md
# App Visium API 

# Installation for dev
- Create a .env from the .env.example
- Run the docker containers
```bash
npm run docker:dev
```

# Run commands of dependencies (Alembic, etc)
```bash
docker exec -it <container-name> <command>
```

# Things that can be improved
- Create CI/CD Pipelines where the project is tested (unit and e2e) and push the Docker image to the registry
- Create a reverse proxy as a load balancer to don't let the app gets exposed by using 0.0.0.0 (specially in prod)
- Use pipenv where all the dependencies are well managed (+ scripts) instead of the traditional requirements.txt
- Instead of having a general preferences/settings table, create also the preferencesGroups table and link it. After that, link this same table with a Options table creating a polymorphic relationship/inheritance so there can be several types of options structures for every section/group of settings. This is better for scaling the arch.
- For bigger projects where there are hundreds of foreign keys, it's better to create different unique table files for them and then, make an order of run.
- Instead of using just 1 JWT, use 2 (Access token and refresh token):
    - The refresh token is stored in db and has the ability to let generate access tokens. It can be revoked manually (during logouts, for example). It is stored in the client side part as a httpOnly secure lax/strict cookie so we can check it on server side automatically and don't put it in danger.
    - The access token is stored if possible in memory so everytime we refresh the page, a new acces token will be requested. This is a short life time token (15 min for example unlike the refresh one, which can be much longer).
- Add a CSRF token with double token system:
    - While we store the refresh token securely in a httponly cookie, there can be problems regarding the CSRF attacks. Using a random string token as CSRF token, we attach it to the header of the request and the server will validate if it corresponds with the one which is stored in a httponly cookie as is sent automatically (Double token system).

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [0.3.0](https://github.com/IvaanTorres/app-visium-api/compare/v0.2.0...v0.3.0) (2023-08-28)


### Features

* **general:** change the welcoming message size preferences ([#21](https://github.com/IvaanTorres/app-visium-api/issues/21)) ([f84a6d5](https://github.com/IvaanTorres/app-visium-api/commit/f84a6d59bc2793efbba019922f6b7cf072672dae))
* **general:** get welcoming message size ([#41](https://github.com/IvaanTorres/app-visium-api/issues/41)) ([da97ad2](https://github.com/IvaanTorres/app-visium-api/commit/da97ad2b43922a1bb41c74595f90eafd3e0f3abf))
* **infos:** get number of login ([#39](https://github.com/IvaanTorres/app-visium-api/issues/39)) ([56bc656](https://github.com/IvaanTorres/app-visium-api/commit/56bc65622201779d7de7483b6cb8b6126b3c11af))
* **locale:** get locale ([#40](https://github.com/IvaanTorres/app-visium-api/issues/40)) ([cab8f02](https://github.com/IvaanTorres/app-visium-api/commit/cab8f02b41919504cd49d74339f2d2c2f83c141a))
* **refresh:** generate new access token ([#25](https://github.com/IvaanTorres/app-visium-api/issues/25)) ([4ce7532](https://github.com/IvaanTorres/app-visium-api/commit/4ce7532d7519b3b3d4b40bbe39a019f206bdac43))
* **settings:** change language preferences ([#20](https://github.com/IvaanTorres/app-visium-api/issues/20)) ([5ddf2ac](https://github.com/IvaanTorres/app-visium-api/commit/5ddf2ac528bf09fa42e354c0a43a3938b5ac19ee))
* **settings:** change username ([#22](https://github.com/IvaanTorres/app-visium-api/issues/22)) ([2480ccf](https://github.com/IvaanTorres/app-visium-api/commit/2480ccfda8239d53063bc64d48a912829032ac6f))


### Style/Clean up

* **arch:** split files in their respective folders ([#47](https://github.com/IvaanTorres/app-visium-api/issues/47)) ([9b522a2](https://github.com/IvaanTorres/app-visium-api/commit/9b522a266b44c1a586d9f22d75d28b69241c9077))
* **function-returns:** arrange the function returns to adapt the object types with the ui ([#45](https://github.com/IvaanTorres/app-visium-api/issues/45)) ([347d050](https://github.com/IvaanTorres/app-visium-api/commit/347d050a3228ff10eaed4313dd3e4cc9a50b5c2f))

## [0.2.0](https://github.com/IvaanTorres/app-visium-api/compare/v0.1.0...v0.2.0) (2023-08-27)


### Features

* **delete-account:** delete users's account ([#23](https://github.com/IvaanTorres/app-visium-api/issues/23)) ([aeee662](https://github.com/IvaanTorres/app-visium-api/commit/aeee662a497392da822c05bf79116c27345705bd))
* **hash:** add salt on hashing ([#16](https://github.com/IvaanTorres/app-visium-api/issues/16)) ([b87d6fe](https://github.com/IvaanTorres/app-visium-api/commit/b87d6fe31e4a4a1ed942c88bf5166e6f249a625b))
* **login:** login user ([#18](https://github.com/IvaanTorres/app-visium-api/issues/18)) ([35f30d7](https://github.com/IvaanTorres/app-visium-api/commit/35f30d74afe2ef6bfd45561a3eac72ce2a0df384))
* **logout:** logout user ([#19](https://github.com/IvaanTorres/app-visium-api/issues/19)) ([f5c0d40](https://github.com/IvaanTorres/app-visium-api/commit/f5c0d401a935c6afc49958da0ba4f84ae3ecb1f4))
* **register:** register new user ([#17](https://github.com/IvaanTorres/app-visium-api/issues/17)) ([8422c79](https://github.com/IvaanTorres/app-visium-api/commit/8422c79e37fbf8a6cb224e3c4fb4625c7d7b4208))
* **security:** create token secrets, password hashing and hmac system based for jwt ([#14](https://github.com/IvaanTorres/app-visium-api/issues/14)) ([af5aaae](https://github.com/IvaanTorres/app-visium-api/commit/af5aaaee67c28ba366670581e78c4b9e706978e3))


### Bug Fixes

* **logout:** get token from authorization and recoke it correctly ([#19](https://github.com/IvaanTorres/app-visium-api/issues/19)) ([c5ef152](https://github.com/IvaanTorres/app-visium-api/commit/c5ef152c1b20b80f7b02b31ef7b2aa6a156dc5d2))

## 0.1.0 (2023-08-24)


### Features

* **boilerplate:** create scaffolding boilerplate ([#1](https://github.com/IvaanTorres/app-visium-api/issues/1)) ([88a371f](https://github.com/IvaanTorres/app-visium-api/commit/88a371f3f9641eb02b08619393542c3e5a806fb2))
* **migrations:** create migrations ([#11](https://github.com/IvaanTorres/app-visium-api/issues/11)) ([c6fa1e1](https://github.com/IvaanTorres/app-visium-api/commit/c6fa1e13e6e2d07ddea0e3912bcc55e863989289))


### Build

* **deps:** install external dependencies API ([#2](https://github.com/IvaanTorres/app-visium-api/issues/2)) ([17f07be](https://github.com/IvaanTorres/app-visium-api/commit/17f07be98f63a86883cc822d04b0ce4e9282027e))
* **docker:** dockerize the app API ([#5](https://github.com/IvaanTorres/app-visium-api/issues/5)) ([24300a3](https://github.com/IvaanTorres/app-visium-api/commit/24300a371c74651e7840b40d2340b96152a2cea8))
* **husky:** install and set up husky API ([#3](https://github.com/IvaanTorres/app-visium-api/issues/3)) ([37ad3db](https://github.com/IvaanTorres/app-visium-api/commit/37ad3db36fdc1655f5d8404fa24cebdca79e9b9a))
* **standard-version:** install and set up standard-version API ([#4](https://github.com/IvaanTorres/app-visium-api/issues/4)) ([137a183](https://github.com/IvaanTorres/app-visium-api/commit/137a183be9fd687aa0a665652747203b18a8509f))


app/api/user.py
from fastapi import APIRouter, Request, HTTPException
from fastapi.responses import JSONResponse
from ..db.config import SessionLocal
from ..models.User import UserModel
from ..db.schemas.User import User
from ..db.schemas.Token import Token
from ..db.schemas.Login import Login
from ..db.schemas.Preference import Preference
import re
from ..shared.helpers.hash import hash, validate_hash, createSalt
from ..shared.helpers.jwt import create_jwt
from ..shared.helpers.date import calculate_future_time
from ..shared.helpers.jwt import check_access
from ..shared.constants.jwt import header

router = APIRouter()

SECRET_KEY = 'my-secret-key'

# ------------------------------- LOGIN SYSTEM ------------------------------- #

# Considering the project requirements, I cannot use Pydantic for input validation and data serialization.
@router.post("/register")
def register(data: dict):
    session = SessionLocal()

    userModel = UserModel(username=data["username"], password=data["password"], email=data["email"])

    # Validate the data
    password_pattern = r"^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[!@#$%^&*])(?=.{8,})"
    user_pattern = r"^[a-zA-Z0-9_.]{3,20}$"
    email_pattern = r"\S+@\S+\.\S+"

    try:
        if not userModel.username:
            raise HTTPException(status_code=400, detail="Username is required")
        if not userModel.password:
            raise HTTPException(status_code=400, detail="Password is required")
        if not userModel.email:
            raise HTTPException(status_code=400, detail="Email is required")
        if session.query(User).filter(User.username == userModel.username).first():
            raise HTTPException(status_code=400, detail="Username already exists")
        if session.query(User).filter(User.email == userModel.email).first():
            raise HTTPException(status_code=400, detail="Email already exists")
        if not re.match(user_pattern, userModel.username):
            raise HTTPException(status_code=400, detail="Username contains invalid characters")
        if not re.match(password_pattern, userModel.password):
            raise HTTPException(status_code=400, detail="Password contains invalid characters")
        if not re.match(email_pattern, userModel.email):
            raise HTTPException(status_code=400, detail="Email contains invalid characters")
        
    except HTTPException as e:
        return {
            "error": {
                "status_code": e.status_code,
                "message": e.detail
            }
        }
    
    # Hash the password
    salt = createSalt(16)
    hashed_password = hash(userModel.password, salt)

    # Set and store the user data, their token, their preferences, and their login count
    user = User(username=userModel.username, password=hashed_password, email=userModel.email)
    session.add(user)
    session.commit()

    stored_user = session.query(User).filter(User.username == userModel.username).first()
    # Generate the secret key for both tokens
    # A good improvement would be to store the secret key in a secure secrets management system as HashiCorp Vault, AWS Secrets Manager, or a secure database
    # user_secret = generate_secret_key()
    # Generate the access token
    access_expiration_time = 1 # 1h
    access_token = create_jwt(
        header, 
        {
            "user_id": stored_user.id, 
            "exp": calculate_future_time(access_expiration_time)
        }, 
        SECRET_KEY
    )

    # Generate the refresh token
    refresh_expiration_time = 24 # 1 day
    refresh_token = create_jwt(
        header, 
        {
            "user_id": stored_user.id, 
            "exp": calculate_future_time(refresh_expiration_time)
        }, 
        SECRET_KEY
    )

    token = Token(token=refresh_token, user=user)
    login = Login(nb_logins=1, user=user)
    preference = Preference(locale=data["locale"], user=user)

    session.add(token)
    session.add(login)
    session.add(preference)
    session.commit()
    session.close()

    return {
        "message": "User registered successfully",
        "data": {
            "access_token": {
                "token": access_token,
                "expires_in": calculate_future_time(access_expiration_time)
            },
            "refresh_token": {
                "token": refresh_token,
                "expires_in": calculate_future_time(refresh_expiration_time)
            }
        }
    }

# Considering the project requirements, I cannot use Pydantic for input validation and data serialization.
@router.post("/login")
def login(data: dict):
    session = SessionLocal()

    if data.get("username"):
        user = UserModel(username=data["username"], password=data["password"])
    elif data.get("email"):
        user = UserModel(email=data["email"], password=data["password"])



    # Validate the data
    password_pattern = r"^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[!@#$%^&*])(?=.{8,})"
    user_pattern = r"^[a-zA-Z0-9_.]{3,20}$"
    email_pattern = r"\S+@\S+\.\S+"

    try:
        if not user.username and not user.email:
            raise HTTPException(status_code=400, detail="Username or email is required")
        if not user.password:
            raise HTTPException(status_code=400, detail="Password is required")
        if user.username and not re.match(user_pattern, user.username):
            raise HTTPException(status_code=400, detail="Username contains invalid characters")
        if not re.match(password_pattern, user.password):
            raise HTTPException(status_code=400, detail="Password contains invalid characters")
        if user.email and not re.match(email_pattern, user.email):
            raise HTTPException(status_code=400, detail="Email contains invalid characters")
        
        # Check if the user or email exists
        print(session.query(User).filter(User.username == user.username).first(), user.email)
        if user.username:
            if session.query(User).filter(User.username == user.username).first() is None:
                print("here")
                raise HTTPException(status_code=400, detail="Username does not exist")
        if user.email:
            if session.query(User).filter(User.email == user.email).first() is None:
                raise HTTPException(status_code=400, detail="Email does not exist")
        
    except HTTPException as e:
        return {
            "error": {
                "status_code": e.status_code,
                "message": e.detail
            }
        }
    
    storedUser = None
    if user.username:
        storedUser = session.query(User).filter(User.username == user.username).first()
    if user.email:
        storedUser = session.query(User).filter(User.email == user.email).first()
    
    # Verify the password
    try: 
        is_valid = validate_hash(user.password, storedUser.password)
        if not is_valid:
            raise HTTPException(status_code=400, detail="Password is invalid")
    except HTTPException as e:
        return {
            "error": {
                "status_code": e.status_code,
                "message": e.detail
            }
        }

    # Generate the secret key for both tokens
    # A good improvement would be to store the secret key in a secure secrets management system as HashiCorp Vault, AWS Secrets Manager, or a secure database
    # user_secret = generate_secret_key()
    # Generate the access token
    access_expiration_time = 1 # 1h
    access_token = create_jwt(
        header, 
        {
            "user_id": storedUser.id, 
            "exp": calculate_future_time(access_expiration_time),
        }, 
        SECRET_KEY
    )

    # Generate the refresh token
    refresh_expiration_time = 24 # 1 day
    refresh_token = create_jwt(
        header, 
        {
            "user_id": storedUser.id, 
            "exp": calculate_future_time(refresh_expiration_time)
        }, 
        SECRET_KEY
    )

    # +1 to the loggin count and add the refresh token
    login = session.query(Login).filter(Login.user_id == storedUser.id).first()
    login.nb_logins += 1

    token = Token(token=refresh_token, user=storedUser)
    
    preference = session.query(Preference).filter(Preference.user_id == storedUser.id).first()
    preference.locale = data["locale"]

    session.add(login)
    session.add(token)
    session.commit()
    session.close()

    return {
        "message": "User logged in successfully",
        "data": {
            "access_token": {
                "token": access_token,
                "expires_in": calculate_future_time(access_expiration_time)
            },
            "refresh_token": {
                "token": refresh_token,
                "expires_in": calculate_future_time(refresh_expiration_time)
            }
        }
    }

# Considering the project requirements, I cannot use Pydantic for input validation and data serialization.
@router.post("/logout")
def logout(request: Request):
    # refresh_token = request.cookies.get("x-refresh-token")
    refresh_token_with_bearer = request.headers["Authorization"]
    refresh_token = refresh_token_with_bearer.split(" ")[1]

    try:
        if not refresh_token:
            raise HTTPException(status_code=400, detail="Token is required")
    except HTTPException as e:
        return {
            "error": {
                "status_code": e.status_code,
                "message": e.detail
            }
        }
    
    # Revoke the token
    try:
        session = SessionLocal()
        storedToken = session.query(Token).filter(Token.token == refresh_token).first()
        if not storedToken:
            raise HTTPException(status_code=400, detail="Token does not exist")

        storedToken.is_revoked = 1

        session.commit()
        session.close()
    except HTTPException as e:
        return {
            "error": {
                "status_code": e.status_code,
                "message": e.detail
            }
        }

    response = JSONResponse(content={
        "message": "Logged out successfully",
        "data": {
            "is_logged_out": True
        }
    })
    # response.delete_cookie(key="x-refresh-token")  # Clear the cookie
    return response

@router.post("/token/refresh")
def refresh_token(request: Request):
    refresh_token_with_bearer = request.headers["Authorization"]
    refresh_token = refresh_token_with_bearer.split(" ")[1]

    # Check if the token is valid
    try:
        if not refresh_token:
            raise HTTPException(status_code=400, detail="Token is required")
    except HTTPException as e:
        return {
            "error": {
                "status_code": e.status_code,
                "message": e.detail
            }
        }
    
    # Revoke the token
    try:
        session = SessionLocal()
        storedToken = session.query(Token).filter(Token.token == refresh_token).first()
        if not storedToken:
            raise HTTPException(status_code=400, detail="Token does not exist")

        storedToken.is_revoked = 1
    except HTTPException as e:
        return {
            "error": {
                "status_code": e.status_code,
                "message": e.detail
            }
        }
    
    # Generate a new token
    try:
        refresh_token_payload = check_access(refresh_token, SECRET_KEY)
        if refresh_token_payload:
            user_id = refresh_token_payload["user_id"]

            # Generate the refresh token
            refresh_expiration_time = 24 # 1 day
            refresh_token = create_jwt(
                header, 
                {
                    "user_id": user_id, 
                    "exp": calculate_future_time(refresh_expiration_time)
                }, 
                SECRET_KEY
            )

            token = Token(token=refresh_token, user_id=user_id)

            session.add(token)
            session.commit()
            session.close()

            response = JSONResponse(content={
                "message": "Token refreshed successfully",
                "data": {
                    "refresh_token": {
                        "token": refresh_token,
                        "expires_in": calculate_future_time(refresh_expiration_time)
                    }
                }
            })
            return response
    except HTTPException as e:
        return {
            "error": {
                "status_code": e.status_code,
                "message": e.detail
            }
        }


# ------------------------------- USER SYSTEM ------------------------------- #
# Considering the project requirements, I cannot use Pydantic for input validation and data serialization.
@router.delete("/delete-account")
def delete_account(request: Request):
    try:
        refresh_token_with_bearer = request.headers["Authorization"]
        refresh_token = refresh_token_with_bearer.split(" ")[1]

        access_token_payload = check_access(refresh_token, SECRET_KEY)

        if access_token_payload:
            user_id = access_token_payload["user_id"]

            session = SessionLocal()
            storedUser = session.query(User).filter(User.id == user_id).first()
            if not storedUser:
                raise HTTPException(status_code=400, detail="User does not exist")

            # Delete tokens (sessions) linked to user
            storedTokens = session.query(Token).filter(Token.user_id == user_id).all()
            for token in storedTokens:
                session.delete(token)

            # Delete the login records linked to user
            storedLogin = session.query(Login).filter(Login.user_id == user_id).first()
            session.delete(storedLogin)

            # Delete the preferences linked to user
            storedPreference = session.query(Preference).filter(Preference.user_id == user_id).first()
            session.delete(storedPreference)

            session.delete(storedUser)
            session.commit()
            session.close()


            return {
                "message": "User deleted successfully",
                "data": {
                    "is_deleted": True
                }
            }
    except HTTPException as e:
        return {
            "error": {
                "status_code": e.status_code,
                "message": e.detail
            }
        }

# Considering the project requirements, I cannot use Pydantic for input validation and data serialization.
@router.get("/user")
def get_user(request: Request):
    try:
        refresh_token_with_bearer = request.headers["Authorization"]
        refresh_token = refresh_token_with_bearer.split(" ")[1]

        refresh_token_payload = check_access(refresh_token, SECRET_KEY)

        if refresh_token_payload:
            user_id = refresh_token_payload["user_id"]

            session = SessionLocal()
            storedUser = session.query(User).filter(User.id == user_id).first()
            if not storedUser:
                raise HTTPException(status_code=400, detail="User does not exist")

            session.close()

            return {
                "message": "User retrieved successfully",
                "data": {
                    "username": storedUser.username,
                    "email": storedUser.email
                }
            }

    except HTTPException as e:
        return {
            "error": {
                "status_code": e.status_code,
                "message": e.detail
            }
        }   
